FA in fv_regional_bc and external_ic

mzhangw opened the pull request:
This PR focus on two topics: 
-  In fv_regional_bc: initialize q_rimef and add the cloud physics auto conversion routine for FA scheme.
 - Fix the ntrac> ntracers limitation in external_ic.
- Add temperature rebalance in the last step of Lagrangian_to_Eulerian step
- Fix a delz bug in subroutine neg_adj2
 - Further optimize the use of q_rimef in fv_dynamic

Full RTs passed except fv3_ccpp_gfdlmp_hwrfsas generated non-bit4bit results due to a known fix in CCPP HWRF physics.

mzhangw committed:
- fix the ntrac> ntracers limitation in external_ic.
  - In fv_regional_bc: initialize q_rimef and add the cloud physics auto conversion routine for FA scheme.

mzhangw committed:
add negative hydrometeors adjustment for F-A scheme

climbfuji commented on model/fv_dynamics.F90:
```diff
@@ -778,8 +782,8 @@ subroutine fv_dynamics(npx, npy, npz, nq_tot,  ng, bdt, consv_te, fill,
        call prt_mxm('ice_wat_dyn', q(isd,jsd,1,ice_wat), is, ie, js, je, ng, npz, 1.,gridstruct%area_64, domain)
-       call prt_mxm('snowwat_dyn', q(isd,jsd,1,snowwat), is, ie, js, je, ng, npz, 1.,gridstruct%area_64, domain)
-       call prt_mxm('graupel_dyn', q(isd,jsd,1,graupel), is, ie, js, je, ng, npz, 1.,gridstruct%area_64, domain)
+       !call prt_mxm('snowwat_dyn', q(isd,jsd,1,snowwat), is, ie, js, je, ng, npz, 1.,gridstruct%area_64, domain)
```
Please use the same method as in lines 730 +/- - check if the index is larger than zero to print, otherwise skip. Don't comment out.  What about q_rimef?

mzhangw replied:
Good point. It reminds me that q_rimef is not  counted in Lagrangian_to_Eulerian step.

mzhangw replied:
BTW, we can not do similar thing here as in line 730, since the first argument of prt_mxm is an arbitrary character and we can not guarantee there is snowwat, graupel or q_rimef to print out in HWRF physics or GFDL MP

climbfuji commented on model/fv_dynamics.F90:
```diff
@@ -909,6 +913,29 @@ subroutine fv_dynamics(npx, npy, npz, nq_tot,  ng, bdt, consv_te, fill,
+       call prt_mxm('liq_wat_dyn', q(isd,jsd,1,liq_wat), is, ie, js, je, ng, npz, 1.,gridstruct%area_64, domain)              
+       call prt_mxm('rainwat_dyn', q(isd,jsd,1,rainwat), is, ie, js, je, ng, npz, 1.,gridstruct%area_64, domain)              
+       call prt_mxm('ice_wat_dyn', q(isd,jsd,1,ice_wat), is, ie, js, je, ng, npz, 1.,gridstruct%area_64, domain)              
+!       call prt_mxm('snowwat_dyn', q(isd,jsd,1,snowwat), is, ie, js, je, ng, npz, 1.,gridstruct%area_64, domain)              
```
Ditto.

mzhangw replied:
I will delete this line since no snowat in HWRF

climbfuji replied:
Ok

climbfuji commented on model/fv_sg.F90:
```diff
@@ -2138,7 +2166,9 @@ subroutine neg_adj2(is, ie, js, je, ng, kbot, hydrostatic,   &
         ql(i,j,k) = ql2(i,j)
         qi(i,j,k) = qi2(i,j)
+        if (present(qs)) then
         qs(i,j,k) = qs2(i,j)
```
adjust indent (add three spaces)

climbfuji commented on tools/external_ic.F90:
```diff
@@ -570,8 +570,9 @@ subroutine get_nggps_ic (Atm, fv_domain, dt_atmos )
-      if (ntrac > ntracers) call mpp_error(FATAL,'==> External_ic::get_nggps_ic: more NGGPS tracers &
-                                 &than defined in field_table '//trim(fn_gfs_ctl)//' for NGGPS IC')
+!mzhang: not in FA
+!      if (ntrac > ntracers) call mpp_error(FATAL,'==> External_ic::get_nggps_ic: more NGGPS tracers &
```
Do we know why this test has been added there? Are there any potentially dangerous consequences if ntrac > ntracers?

mzhangw replied:
The max number of NGGPS tracers (ntrac) is predefined in gfs_ctrl.nc as 7. I guess that it was the max tracers supported at that time?

climbfuji replied:
Yeah, but why did they put the requirement there? Is there anything wrong if the number of tracers in the file is larger than what is used? I don't see what should be wrong, but you never know if there is some code elsewhere that relies on this condition. But maybe other code changes have made this check superfluous. Let's comment it out as you suggested and see what GFDL says when we want to merge it back.

ChunxiZhang-NOAA replied:
I don't know if the modification here will impact other places either.

ChunxiZhang-NOAA replied:
!zhang nwat=4
    if ( Atm%flagstruct%nwat .eq. 4 ) then
       do k=1,npz
            BC_side%q_BC(i,j,k,rainwat) = 0.
             BC_side%q_BC(i,j,k,q_rimef) = 0.
             BC_side%q_BC(i,j,k,snowwat) = 0.
             BC_side%q_BC(i,j,k,graupel) = 0.
                  .......
     q_BC is defined by (:,:,:,1:ntracers). So I don't think snowwat and graupel need to be included here.
     I am not sure if mp_auto_conversion_fa is necessary.

mzhangw committed:
- add temperature rebalance in the last step of lagrangian_to_Eulerian step
 - fix a delz bug in subroutine neg_adj2
 - further optimize the use of q_rimef in fv_dynamic

climbfuji reviewed (commented):
I am ok with these changes, but I would like to ask @ChunxiZhang-NOAA to look at this and approve, too. We will need PRs for fv3atm and ufs-weather-model for the submodule pointer updates; @mzhangw please create those to prepare for committing the code. Thanks!

mzhangw commented:
Done.
https://github.com/NCAR/ufs-weather-model/pull/63


> On Aug 5, 2020, at 4:31 PM, Dom Heinzeller <notifications@github.com> wrote:
> 
> 
> @climbfuji commented on this pull request.
> 
> I am ok with these changes, but I would like to ask @ChunxiZhang-NOAA <https://github.com/ChunxiZhang-NOAA> to look at this and approve, too. We will need PRs for fv3atm and ufs-weather-model for the submodule pointer updates; @mzhangw <https://github.com/mzhangw> please create those to prepare for committing the code. Thanks!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/NCAR/GFDL_atmos_cubed_sphere/pull/17#pullrequestreview-462068165>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AG7TW2WDGL6TDZ57PWU6RALR7HMVZANCNFSM4PQ4JI7Q>.
>

mzhangw commented:
For now, it does not impact all existing RTs and makes HWRF physics working with ntrac = 6 in field_table (no cld_amt).

Man 
> On Aug 6, 2020, at 9:12 AM, ChunxiZhang-NOAA <notifications@github.com> wrote:
> 
> 
> @ChunxiZhang-NOAA commented on this pull request.
> 
> In tools/external_ic.F90 <https://github.com/NCAR/GFDL_atmos_cubed_sphere/pull/17#discussion_r466484877>:
> 
> > @@ -570,8 +570,9 @@ subroutine get_nggps_ic (Atm, fv_domain, dt_atmos )
>  
>  !--- read in the number of tracers in the NCEP NGGPS ICs
>        call read_data ('INPUT/'//trim(fn_gfs_ctl), 'ntrac', ntrac, no_domain=.TRUE.)
> -      if (ntrac > ntracers) call mpp_error(FATAL,'==> External_ic::get_nggps_ic: more NGGPS tracers &
> -                                 &than defined in field_table '//trim(fn_gfs_ctl)//' for NGGPS IC')
> +!mzhang: not in FA
> +!      if (ntrac > ntracers) call mpp_error(FATAL,'==> External_ic::get_nggps_ic: more NGGPS tracers &
> I don't know if the modification here will impact other places either.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/NCAR/GFDL_atmos_cubed_sphere/pull/17#discussion_r466484877>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AG7TW2TWW3JS2HU7BCZYST3R7LB47ANCNFSM4PQ4JI7Q>.
>

ChunxiZhang-NOAA reviewed (commented):
for fv_regional_bc.F90

!zhang nwat=4
if ( Atm%flagstruct%nwat .eq. 4 ) then
do k=1,npz
BC_side%q_BC(i,j,k,rainwat) = 0.
BC_side%q_BC(i,j,k,q_rimef) = 0.
BC_side%q_BC(i,j,k,snowwat) = 0.
BC_side%q_BC(i,j,k,graupel) = 0.
.......
**q_BC is defined by (:,:,:,1:ntracers). So I don't think snowwat and graupel need to be included here.
I am not sure if mp_auto_conversion_fa is necessary.**

mzhangw commented:
good catch!

> On Aug 6, 2020, at 9:49 AM, ChunxiZhang-NOAA <notifications@github.com> wrote:
> 
> 
> @ChunxiZhang-NOAA commented on this pull request.
> 
> for fv_regional_bc.F90
> 
> !zhang nwat=4
> if ( Atm%flagstruct%nwat .eq. 4 ) then
> do k=1,npz
> BC_side%q_BC(i,j,k,rainwat) = 0.
> BC_side%q_BC(i,j,k,q_rimef) = 0.
> BC_side%q_BC(i,j,k,snowwat) = 0.
> BC_side%q_BC(i,j,k,graupel) = 0.
> .......
> q_BC is defined by (:,:,:,1:ntracers). So I don't think snowwat and graupel need to be included here.
> I am not sure if mp_auto_conversion_fa is necessary.
> 
> In tools/external_ic.F90 <https://github.com/NCAR/GFDL_atmos_cubed_sphere/pull/17#discussion_r466504609>:
> 
> > @@ -570,8 +570,9 @@ subroutine get_nggps_ic (Atm, fv_domain, dt_atmos )
>  
>  !--- read in the number of tracers in the NCEP NGGPS ICs
>        call read_data ('INPUT/'//trim(fn_gfs_ctl), 'ntrac', ntrac, no_domain=.TRUE.)
> -      if (ntrac > ntracers) call mpp_error(FATAL,'==> External_ic::get_nggps_ic: more NGGPS tracers &
> -                                 &than defined in field_table '//trim(fn_gfs_ctl)//' for NGGPS IC')
> +!mzhang: not in FA
> +!      if (ntrac > ntracers) call mpp_error(FATAL,'==> External_ic::get_nggps_ic: more NGGPS tracers &
> !zhang nwat=4
> if ( Atm%flagstruct%nwat .eq. 4 ) then
> do k=1,npz
> BC_side%q_BC(i,j,k,rainwat) = 0.
> BC_side%q_BC(i,j,k,q_rimef) = 0.
> BC_side%q_BC(i,j,k,snowwat) = 0.
> BC_side%q_BC(i,j,k,graupel) = 0.
> .......
> q_BC is defined by (:,:,:,1:ntracers). So I don't think snowwat and graupel need to be included here.
> I am not sure if mp_auto_conversion_fa is necessary.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/NCAR/GFDL_atmos_cubed_sphere/pull/17#pullrequestreview-462640680>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AG7TW2TNO5VYLIMCLNRBLVLR7LGIHANCNFSM4PQ4JI7Q>.
>

climbfuji commented:
> for fv_regional_bc.F90
> 
> !zhang nwat=4
> if ( Atm%flagstruct%nwat .eq. 4 ) then
> do k=1,npz
> BC_side%q_BC(i,j,k,rainwat) = 0.
> BC_side%q_BC(i,j,k,q_rimef) = 0.
> BC_side%q_BC(i,j,k,snowwat) = 0.
> BC_side%q_BC(i,j,k,graupel) = 0.
> .......
> **q_BC is defined by (:,:,:,1:ntracers). So I don't think snowwat and graupel need to be included here. I am not sure if mp_auto_conversion_fa is necessary.**

@ChunxiZhang-NOAA please submit a follow-up PR to correct whatever needs to be corrected. Thanks!
